oak-run/src/main/java/org/apache/jackrabbit/oak/run/Downloader.java

```diff
@@ -0,0 +1,263 @@
+
+        private final Item item;
+
+        public DownloaderWorker(Item item) {
+            this.item = item;
+        }
+
+        @Override
+        public ItemResponse call() throws Exception {
+            long t0 = System.nanoTime();
+
+            URLConnection sourceUrl = new URL(item.source).openConnection();
+            sourceUrl.setConnectTimeout(Downloader.this.connectTimeoutMs);
+            sourceUrl.setReadTimeout(Downloader.this.readTimeoutMs);
+
+            Path destinationPath = Paths.get(item.destination);
+            Files.createDirectories(destinationPath.getParent());
+            long size;
+            try (ReadableByteChannel byteChannel = Channels.newChannel(sourceUrl.getInputStream());
+                 FileOutputStream outputStream = new FileOutputStream(destinationPath.toFile())) {
+                size = outputStream.getChannel()
+                        .transferFrom(byteChannel, 0, Long.MAX_VALUE);
+            }
+            long elapsed = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - t0);
+            if (slowLogThreshold > 0 && elapsed >= slowLogThreshold) {
+                LOG.warn("{} [{}] downloaded in {} ms", item.source, IOUtils.humanReadableByteCount(size), elapsed);
+            } else {
+                LOG.debug("{} [{}] downloaded in {} ms", item.source, IOUtils.humanReadableByteCount(size), elapsed);
+            }
+            return ItemResponse.success(size);
+        }
+
+        @Override
+        public String toString() {
+            return "DownloaderWorker{" +
+                    "item=" + item +
+                    '}';
+        }
+    }
+
+    private static class RetryingCallable<V> implements Callable<V> {
+        private final Callable<V> callable;
+        private final int retries;
+
+        public RetryingCallable(int retries, Callable<V> callable) {
+            this.retries = retries;
+            this.callable = callable;
+        }
+
+        public V call() {
+            int retried = 0;
+            // Save exceptions that are thrown after each failure, so they can be printed if all retries fail
+            List<Throwable> exceptions = new ArrayList<>();
+
+            // Loop until it doesn't throw an exception or max number of tries is reached
+            while (true) {
+                try {
+                    return callable.call();
+                } catch (Exception e) {
+                    retried++;
+                    exceptions.add(e);
+
+                    // Throw exception if number of tries has been reached
+                    if (retried == retries) {
+                        // Get a string of all exceptions that were thrown
+                        StringBuilder exceptionsString = new StringBuilder();
+                        for (int i = 0; i < exceptions.size(); i++) {
+                            exceptionsString.append("\nFailure ").append(i + 1).append(": ").append(exceptions.get(i));
+                        }
```

Comment:
This might be a massive string, mostly with repeated exceptions. In the worst case scenario, we will do 1000 retries for an unlimited number of blobs. Let's say 1000 blobs. If all the attempts fail with an exception, this code will generate a message containing 1_000_000 exception stack traces. There should be a limit on the size of this message. Or try to group the exceptions by type, and log something like

```
 30x:  IOException - Connection timeout....
```

Reply:
To be precise, this generates a string for each blob and not a huge single string with all the blobs that failed. Moreover, only the exception messages are printed and not the stack traces. Only the last exception stack trace is reported.

I have improved the logs, following your suggestion to remove duplicate messages.